### PR TITLE
Signatures use JSDoc to determine optionality

### LINKED ISSUE #9228: Parameters don't use JSDoc optionality
_From @jdforsythe on June 16, 2016 20:46_

- VSCode Version: 1.2.1
- OS Version: Windows 10

Steps to Reproduce:

**proto.js**
```js
module.exports = MyClass;

function MyClass() {
  this.prop = null;
}

/**
 * @param  {any} required
 * @param  {any} [notRequired]
 * @returns {MyClass}
 */
MyClass.prototype.optionalParam = function(required, notRequired) {
  this.prop = required;

  if (notRequired)
    this.prop += notRequired;

  return this;
};

MyClass.prototype.secondMethod = function(param) {
  return 'Hello';
};
```

**index.js**
```js
const proto = require('./proto');

let pInst = new proto();

//pInst.optionalParam('hello')
//pInst.optionalParam('hello', null)
```

Uncomment the second line and hit `.` after it. You'll get the proper Intellisense.

properintelli

Uncomment the first line and hit `.`.

badintelli

The Intellisense is broken if you don't supply something for the parameter, even though it is clearly recognized as optional.

optional

_Copied from original issue: Microsoft/vscode#7793_

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #9228

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -4414,8 +4414,10 @@ namespace ts {
                     }
                 }
             }
+        }
 
-            if (hasQuestionToken(node)) {
+        function isOptionalParameter(node: ParameterDeclaration) {
+            if (isJSDocOptionalParameter(node) || hasQuestionToken(node)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Swap this check around so that the TS check is faster for the positive case.

Comment (sandersn) [inline src/compiler/checker.ts]: Maybe. Isn't `flags & NodeFlags.JavaScriptFile` faster than `switch(node.kind)` with 7 cases? In other words, I think the speed gain will be absolutely tiny. And the code was originally written in this order, although that's (apparently) not meaningful.

Comment (sandersn) [inline src/compiler/checker.ts]: Regardless of performance, I think I like the way it reads better to flip it.

**Comment by mhegazy** [comment]
👍

**Pull request merged**